GPII-2927:  Fixed code coverage for process reporter components.

klown opened the pull request:
@the-t-in-rtf I have modified the HTML test fixtures as you suggested.  Code coverage is much improved.  Do you have time to review?  Thanks!

the-t-in-rtf commented:
@klown, I don't have permission to assign things or to perform a merge, but I'm happy to leave comments.

the-t-in-rtf commented:
@klown, I wanted to raise some points about the code coverage results, but can't link to line numbers because nyc isn't new enough.   I'd suggest running npm outdated while you're there and at a minimum updating nyc, testem, and gpii-testem to the latest released versions.  I did this locally, and saved a copy of the report here:

https://the-t-in-rtf.github.io/coverage-reports/universal/20180405/index.html

I wanted to ask you about a couple of branches that aren't reported as tested.  In particular, I wanted to ask you about this one:

https://the-t-in-rtf.github.io/coverage-reports/universal/20180405/universal/gpii/node_modules/processReporter/src/processesBridge.js.html#L186

For whatever reason a process isn't returned, IMO you should be able to confirm that the system handles that appropriately.  It may be that another fixture needs to be set up to report coverage, or that there simply aren't tests.  If there aren't tests, let's talk through how hard it would be to add them, hopefully it's the kind of thing you can take care of in passing?

the-t-in-rtf commented:
Also, we need to tighten up the "nonSources" option in gpii-testem, things like these shouldn't be included in the report:

https://the-t-in-rtf.github.io/coverage-reports/universal/20180405/universal/gpii/node_modules/transformer/test/js/index.html
https://the-t-in-rtf.github.io/coverage-reports/universal/20180405/universal/gpii/node_modules/settingsHandlers/test/web/js/index.html

I'll try generating "node only" and "web only" reports to see who's creating those, depending on the results, we'll either need to update the NYC config or the gpii-testem component options.  More on that in a bit.

the-t-in-rtf commented:
OK, the node tests are fine, it's clearly the web tests that bring in the test content.  I'll suggest a fix shortly.

the-t-in-rtf commented:
If you add the following to the `nonSources` option in `testem.js`, the extraneous test content will be excluded:

```
            "./**/*Tests.js",
            "./**/*TestEnv.js",
            "./**/*TestsUtils.js",
```

klown commented:
@the-t-in-rtf Thanks for the comments!  I think I have addressed them with the changes I've made -- have a look.

Specifics:

> I wanted to ask you about a couple of branches that aren't reported as tested. In particular, I wanted to ask you about this one:
> ...
> hopefully it's the kind of thing you can take care of in passing?

No, it's simple to add a test for that, and the other lines that were marked in red.  I'm not sure about the lines marked with "Else path not taken", since there is no "else" path to take.

> If you add the following to the nonSources option in testem.js, the extraneous test content will be excluded:
> 
>  ```
>            "./**/*Tests.js",
>             "./**/*TestEnv.js",
>             "./**/*TestsUtils.js",
> 
> ```

Sure, but isn't that a broader, global issue than improving coverage for the process reporter components?  Perhaps another JIRA?

klown commented:
Hi @the-t-in-rtf,

> I understand what you mean, but I often am asked to (and take care of) small stuff that is unrelated to my immediate work as part of an active PR, and I tend to think it's fair game to fix small broken windows when we notice them.

Fair enough:  I've created [GPII-2936](https://issues.gpii.net/browse/GPII-2936), and added the to-be-excluded-files to `testem.js` as you specified above.

> Whether or not you have a literal else block, whenever you have an if which is always hit, the system correctly reports that you have not ever confirmed that the system works without traversing that if block.

D'oh!  My question was literally and simply (and naively), "there is no else-block to check so how can I write a test for it?".  But, as you noted, there is an _implicit_ else in these cases.  Of course there is.  Chalk it up to it being late on a Friday afternoon...  And, that's why you have reviewers, folks.  Anyhow, I have added coverage for all the implicit else-blocks listed in the report.

> In the case I shared above, I think it's important to talk about why we allow that optional argument, and if we agree it's useful, to add tests to cover the implicit else.

Here's the rationale: for an actual OS*, it's relatively expensive to get the list of processes -- the `procArray` -- so, if you just generated it, you can use it again.  On the other hand, the list becomes stale relatively quickly, so if you think the list is out-of-date, don't pass it in.  It is left to the client to determine the trade-off between expense and accuracy.

*- in these (universal) tests, the list is a mock and in memory, and getProcessList() is blazing fast.

New version coming up...

the-t-in-rtf commented:
Ugh, the last passing CI build was actually a failure with a bunch of NullPointerException messages.  Confirming locally and checking out the updated coverage report.

klown commented:
Thanks @the-t-in-rtf 

> Ugh, the last passing CI build was actually a failure with a bunch of NullPointerException messages. 

Weird.  As you said, there are NullPointerExceptions in the [general console log](https://ci.gpii.net/job/universal-tests/811/console), but the logs for the [browser](https://ci.gpii.net/job/universal-browser-tests/805/console) and [node](https://ci.gpii.net/job/universal-browser-tests/805/console) tests show that all tests were run and passed (32 and 1256, respectively). Indeed the other console logs show that everything is tickety-boo. CI is schizoid.

I'm going to look more closely at the process reporter tests in the windows and linux branches for similar coverage issues that the work here revealed.   I'll be using the tools in your [GPII-2493 pull](https://github.com/GPII/windows/pull/131).

Thanks again.

the-t-in-rtf commented:
Hi, @klown.  Sorry about the timing here, looks like the underlying branch was just merged with master.  Probably it's best to close this and submit a (linked) PR against the GPII repo.

klown commented:
Hi @the-t-in-rtf .  You mean that [your GPII-2483 branch/pull](https://github.com/GPII/windows/pull/131) for code coverage in *Windows* was merged.  The pull request *here* (GPII-2927) is for Universal and is based its master branch.   Still, it's a bit out of date, so I've merged in the latest from universal master, and pushed.

(And, I'll go rebase [my Windows pull request](https://github.com/the-t-in-rtf/windows/pull/1) as you suggest).
